### PR TITLE
Handle function template default type arguments

### PR DESCRIPTION
**Pull request by bolshakov-a**
Besides adding default type arguments of function templates into `resugar_map`, this PR introduces short-circuiting of handling a function reference when such a reference occur due to calling that function. Previously, that handling was just redundant. But after turning default argument analysis on, it started causing a bug with unwanted default type template argument reporting when it is actually overridden by the deduced one.

The problem arises because there is a lack of the full context for `resugar_map` and provided type set determination: actual function call arguments are not considered on function reference analyzing. Hence, a default type argument of a function template may be placed into `resugar_map` as defaulted despite being overridden by some of function call argument types.

### CODE REVIEW

File: iwyu_ast_util.cc [hunk 1]
```diff
@@ -888,6 +890,29 @@ static const Type* GetSugaredTypeOf(const Expr* expr) {
   return v.sugared.getTypePtr();
 }
 
+static map<const Type*, const Type*> GetDefaultedArgResugarMap(
+    const FunctionDecl* decl) {
+  map<const Type*, const Type*> res;
+  const FunctionTemplateDecl* template_decl = decl->getPrimaryTemplate();
+  if (!template_decl)
+    return map<const Type*, const Type*>{};
```
Comment (kimgr) [inline iwyu_ast_util.cc:898]: `return res;`

**Review by kimgr — COMMENTED** [review]
This looks like it might be related and might be fixed after these changes: https://github.com/include-what-you-use/include-what-you-use/issues/1121

**Comment by kimgr** [comment]
> This looks like it might be related and might be fixed after these changes: https://github.com/include-what-you-use/include-what-you-use/issues/1121

I meant to ask: is there a testcase there you could salvage? The tests you have look good, but there might be corner cases in @graveljp's suite.

**Comment by bolshakov-a** [comment]
The problems described in #1121 occur to be different from this one, despite the second one is related.

File: tests/cxx/default_tpl_arg.cc [hunk 2]
```diff
@@ -43,6 +43,25 @@ struct Outer2 {
 // IWYU: IndirectTemplate needs a declaration
 Outer2<int, IndirectTemplate<int>> o2;
 
+void Fn() {
+  // IWYU: FnWithDefaultTplArg is...*default_tpl_arg-i1.h
+  // IWYU: IndirectClass is...*indirect.h
```
Comment (kimgr) [inline tests/cxx/default_tpl_arg.cc:48]: So are these attributed here because of the reexport hint (`class IndirectClass;`) before the definition? I think it would be nice to show the "default" behavior as well, where `IndirectClass` should be attributed to the template file... right?

I was playing with graveljp's example here: https://github.com/graveljp/iwyu-bugs/tree/main/default_template_param, but it still reports the default template type at point of instantiation.

Comment (bolshakov-a) [reply tests/cxx/default_tpl_arg.cc:48]: > I was playing with graveljp's example here: https://github.com/graveljp/iwyu-bugs/tree/main/default_template_param, but it still reports the default template type at point of instantiation.

Fixed in #1403. Probably, the test case there is sufficient? Or there should be some without normal default arguments, just with template default arguments?

Comment (kimgr) [reply tests/cxx/default_tpl_arg.cc:48]: Could you add a comment in the test explaining why `IndirectClass` is required at point of instantiation? It looks wrong to me, unless it's due to the forward-decl near definition.

Once I understand that I'll be in a better position to think about more tests :)

File: iwyu.cc [hunk 3]
```diff
@@ -798,6 +799,11 @@ class BaseAstVisitor : public RecursiveASTVisitor<Derived> {
     if (FunctionDecl* fn_decl = DynCastFrom(expr->getDecl())) {
       if (!IsImplicitlyInstantiatedDfn(fn_decl))
         return true;
+      if (const auto* call_expr = current_ast_node_->GetAncestorAs<CallExpr>(
```
Comment (kimgr) [inline iwyu.cc:802]: The formatting on this block is pretty horrible. Seems hard to shorten it, but can you trick `clang-format` by making it longer? E.g. use `current_ast_node()` instead of the raw member?

You could also consider moving the comment line down into the body of the if, but I don't think that alone helps with the wrapping.

File: tests/cxx/default_tpl_arg.cc [hunk 4]
```diff
@@ -43,6 +44,41 @@ struct Outer2 {
...
+  using ProvidingAlias = IndirectClass;
+  ProvidingAlias* p = 0;
+  // IWYU: NonProvidingAlias is...*default_tpl_arg-i1.h
+  NonProvidingAlias* n = 0;
+
+  // IWYU: FnWithNonProvidedDefaultTplArgAndDefaultCallArg is...*default_tpl_arg-i1.h
+  // IWYU: IndirectClass is...*indirect.h
+  FnWithNonProvidedDefaultTplArgAndDefaultCallArg();
+  // TODO: should not report here.
```
Comment (bolshakov-a) [inline tests/cxx/default_tpl_arg.cc:67]: Fixed in #1403. The separate test cases there become probably redundant.

File: tests/cxx/default_tpl_arg.cc (hunk 2 above)
Comment (bolshakov-a) [reply tests/cxx/default_tpl_arg.cc:49]: In fact, instantiation-side analysis doesn't consider forward declarations currently. That's the reason for redundancy on the first IWYU pass when template definition header neither `#include`s nor forward-declares default template argument type. It is suggested for full type info inclusion on the both sides.

**Comment by kimgr** [comment]
@bolshakov-a Yeah, this LGTM. Are you ready to merge?

**Comment by bolshakov-a** [comment]
I think so.
